### ISSUE TITLE
"Short circuiting" in base64's b64decode, decode, decodebytes

### ISSUE THREAD

**Issue by bab888e8-1d2f-4584-b957-5208502b0d1e** [python/cpython#79013]
BPO | [34832](https://bugs.python.org/issue34832)
--- | :---
Nosy | @fbidu, @pwmichaelharris

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-09-28.12:20:13.340>
labels = ['type-bug', 'library']
title = '"Short circuiting" in base64\'s b64decode, decode, decodebytes'
updated_at = <Date 2018-09-29.16:05:02.898>
user = 'https://github.com/pwmichaelharris'
```

bugs.python.org fields:
```python
activity = <Date 2018-09-29.16:05:02.898>
actor = 'fbidu'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-09-28.12:20:13.340>
creator = 'pw.michael.harris'
dependencies = []
files = []
hgrepos = []
issue_num = 34832
keywords = []
message_count = 4.0
messages = ['326630', '326661', '326662', '326680']
nosy_count = 2.0
nosy_names = ['fbidu', 'pw.michael.harris']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue34832'
versions = ['Python 3.5']
```

</p></details>

**Comment by simon-friedberger** [comment]
@fbidu, I agree with your reasoning. The default behavior should error on any non-base-alphabet characters. `validate=True` should be the default. I would even suggest removing the option.

There are inconsistencies in the API, [`b64_decode`](https://docs.python.org/3/library/base64.html#base64.b64decode) has `validate`(default False) but [`urlsafe_b64decode`](https://docs.python.org/3/library/base64.html#base64.urlsafe_b64decode) and `b32_decode` do not. `b32_decode` implicitly has `validate=True` but has a switch for allowing lower-case characters.

A single corrupt character will silently corrupt the rest of the data. Disagreement about which alphabet to use (urlsafe?) will lead to data that decodes fine but isn't the right data.

Since it's fairly trivial (something like `re.sub(r'[a-zA-Z0-9]', ' ', s)`) to remove unwanted characters people who want that could just do that.

**Comment by serhiy-storchaka** [comment]
The originally reported issue was fixed in #145264. Base64 decoder has now the _ignorechars_ parameter which allows finer control on what characters to ignore (see #143214). If '=' is included in _ignorechars_ or in non-strict mode (default), then '=' in the middle of string and excess padding will be ignored, otherwise they will be errors. #73613 added new parameter _padded_ which controls whether the padding is required.

As for `urlsafe_b64decode()`, it has less options, but there are plans to make it stricter by default. Use `b64decode()` with `altchars=b'-_'` if you need more flexibility.
